components/omega/doc/design/Config.md

```diff
@@ -0,0 +1,410 @@
+we use simple overloading to keep a cleaner interface. If the variable
+or config is missing, these functions will print an error message and
+return with a non-zero error argument.
+
+Another interface will allow the setting of a default value if the
+variable is missing from the input config.  This interface simply
+adds the default value as an additional argument, for example:
+
+    refWidth = ConfigGet(hmixConfig, "hmixRefWidth", defaultVal, iErr);
+
+In this case, if the variable does not exist, it will not only
+use the default value but print a warning that the default is
+being used because the entry is missing.
+
+
+#### 4.2.3 Change an existing value
+
+While the intent is for all config variables to be set using the config
+file read interface above, the capability modify a value is also
+required. The syntax is essentially the inverse of the get/retrieval 
+above. Similar to that case, the sub-group will need to be retrieved first.
+
+    Config hmixConfig = ConfigGet(omegaConfig, "hmix", iErr);
+    ConfigSet(hmixConfig, "hmixRefWidth", 10.0e3, iErr);
+    ConfigSet(hmixConfig, "hmixUseRefWidth", true, iErr);
+
+There will be overloaded interfaces for each supported type. For 
+literals (as in the example above), they will be cast to an appropriate
+type according to C++ default type conversion and will be converted
+to the desired type on retrieval (YAML internal storage is ignorant of
+the type and only performs the type cast on retrieval).
+
+#### 4.2.3 Adding new entries
+
+It may be necessary to build up a configuration that does not yet
+exist or add new entries to an existing group. We provide an Add
+interface to distinguish this case from the Set case above.
+
+    // For an existing subgroup:
+    Config hmixConfig = ConfigGet(omegaConfig, "hmix", iErr);
+    ConfigAdd(hmixConfig, "hmixRefWidth", 10.0e3, iErr);
+    ConfigAdd(hmixConfig, "hmixUseRefWidth", true, iErr);
+
+    // To add a new subgroup:
+    Config hmixConfig; // empty Config constructor
+    ConfigAdd(hmixConfig, "hmixRefWidth", 10.0e3, iErr); // build subgroup
+    ConfigAdd(hmixConfig, "hmixUseRefWidth", true, iErr);
+    ConfigAdd(omegaConfig, hmixConfig); // add new subgroup to parent
+
+There will be overloaded interfaces for each supported type. For 
+literals (as in the example above), they will be cast to an appropriate
+type according to C++ default type conversion and will be converted
+to the desired type on retrieval (YAML internal storage is ignorant of
+the type and only performs the type cast on retrieval).
+
+#### 4.2.4 Existence
+
+It is not expected that a user would test the existence since the
+Get/Set functions will perform the test internally. However, to 
+satisfy requirement 2.11, YAML provides a means for such testing.
+The reference to a YAML entry itself returns a status and that is
+used as an existence test. Using the hmix example again:
+
+    if (hmixConfig["hmixRefWidth"]) {
```

Comment:
To me, hmixConfig["hmixRefWidth"] looks to get a value from hmixConfig map instead of checking the existence of the key of "hmixRefWidth". As you noted below, I give plus one to use "ConfigExists" for consistency to other interface.

Reply:
That's the expected way to check in YAML, but I agree that an Exists function would be better. I'll modify the doc accordingly